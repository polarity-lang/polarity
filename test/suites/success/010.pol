codata Fun(a b: Type) {
    Fun(a, b).ap(a: Type, b: Type, x: a) : b
}

data HasType(e: Exp) {
    TApp(e: Exp): HasType(App(e)),
}

data Exp {
    App(e: Exp)
}

data Eval(e1 e2: Exp) {
    EBeta(e: Exp): Eval(App(e), e.id),
}

def Exp.id: Exp {
    App(e) => App(e),
}

data Top { Unit }

def (self: Exp).preservation(e2: Exp): Fun(Eval(self, e2), HasType(e2))
{
    App(e) => \h_eval.
        h_eval.match {
            EBeta(f) => ? : HasType(e.id)
        }
}
