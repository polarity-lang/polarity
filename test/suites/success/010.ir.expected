def .id { App(e) => App(e) }

def self.preservation(e2) {
    App(e) => comatch { .ap(h_eval) => h_eval.match { EBeta(f) => panic!("not yet implemented") } }
}
